libraries/chain/controller.cpp

```diff
@@ -3198,15 +3265,39 @@ struct controller_impl {
                report_block_header_diff(*b, ab.header());
 
                // this implicitly asserts that all header fields (less the signature) are identical
-               EOS_ASSERT(producer_block_id == ab.id(), block_validate_exception, "Block ID does not match",
+               EOS_ASSERT(producer_block_id == ab.id(), block_validate_exception, "Block ID does not match, ${producer_block_id} != ${validator_block_id}",
                           ("producer_block_id", producer_block_id)("validator_block_id", ab.id()));
             }
 
+            // verify received finality digest in action_mroot is the same as the actual one
+            if constexpr (std::is_same_v<BSP, block_state_ptr>) {
```

Comment:
Can you do that check in the above `if constexpr` after the `assemble_block(true, extract_qc_data(b), bsp);`>

Reply:
In earlier version data from assembled_block was required for the check. Not any more. Can move up to save an `if constexpr` and look nicer. Thanks.